### ISSUE TITLE
With filtering, page size counts unfiltered rows. Should be filtered rows?

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#12102]
The Cassandra driver documentation for different languages suggests that the `fetch_size` parameter of a scan "_controls how many rows will be fetched per page._". It is implied - but not explicitly guaranteed in any document that I could find - that if `fetch_size=100`, the scan returns pages of 100 rows each - only returning less in the last page.

In this issue we look at the case of **filtering**, i.e., adding to the scan's WHERE filter like `... and x = 1 ALLOW FILTERING`. In this case, not all rows read during the scan pass the filter and returned to the user. In this case, there is a dilemma:
1. Should `fetch_size=100` mean that Scylla should read 100 rows from disk and return how many of them passed the filter (this can be less than 100, potentially even zero!),
2. Or should `fetch_size=100` mean that Scylla should continue reading until 100 rows are available to be returned, after filtering?

It turns out that the first interpretation is what Scylla does today. I would argue that this interpretation makes more sense:  If the filter is extremely selective and most rows are filtered out, trying to find 100 rows that passed the filter can force Scylla to read a huge number of rows from disk, which can take an unbounded amount of time and lead to time outs. By the way, even when there is no filtering we have the same problems when the data have a lot of tombstones, and there too we may want to return pages smaller than `fetch_size`. This interpretation is also explicitly the one which DynamoDB uses: There it is explicitly specified in documentation that the page size determines the amount of items read from disk before filtering - not the number of items which will be returned.

However, Cassandra went with interpretation 2: If `fetch_size=100` is used, 100 rows are returned even if more had to be read just to find 100 that survived the filtering.  I'm not sure this is guaranteed in every extreme case (such as described above) but in "normal" cases users can use `fetch_size=100` and expect to see 100 results, not fewer.

The reason I'm opening this issue is that we had several users complaining about this incompatibility with Cassandra. One is #11340 which noticed that in Scylla, cqlsh (which uses `fetch_size=100`) sometimes shows empty pages when all 100 rows were filtered out. Another user reported they want to see the first N results from a scan that included filtering, and could not use `LIMIT` because of another bug (#10649) and if they use `fetch_size` they usually get fewer than N results because of the pre-filtering counting.

Here is a cql-pytest which reproduces this issue. It scans a table where half of the rows match the filter, and asks for `fetch_size=3`. In Cassandra, the returned page contains exactly three row and passes the test. In Scylla, it contains just one row (out of the first three rows fetched, just one passed the filter). This result is the same regardless of whether or not we use a secondary index to drive this query (so it's not a secondary-index-specific problem like #10649).

```python
@pytest.mark.parametrize("use_index", [True, False])
def test_filter_and_fetch_size(cql, test_keyspace, use_index, driver_bug_1):
    with new_test_table(cql, test_keyspace, 'pk int primary key, x int, y int') as table:
        if use_index:
            cql.execute(f'CREATE INDEX ON {table}(x)')
        stmt = cql.prepare(f'INSERT INTO {table} (pk, x, y) VALUES (?, ?, ?)')
        cql.execute(stmt, [0, 1, 0])
        cql.execute(stmt, [1, 1, 1])
        cql.execute(stmt, [2, 1, 0])
        cql.execute(stmt, [3, 1, 1])
        cql.execute(stmt, [4, 1, 0])
        cql.execute(stmt, [5, 1, 1])
        cql.execute(stmt, [6, 1, 0])
        cql.execute(stmt, [7, 1, 1])
        # Fetch a filtered page with fetch_size=3 and expect to see 3 rows
        # in the post-filtering results.
        s = cql.prepare(f'SELECT pk FROM {table} WHERE x = 1 AND y = 0 ALLOW FILTERING')
        s.fetch_size = 3
        results = cql.execute(s)
        assert len(results.current_rows) == 3
```

**Comment by nyh** [comment]
Added in pull request https://github.com/scylladb/scylladb/pull/12103 a test reproducing this difference between Cassandra and Scylla: `test_filtering.py::test_filter_and_fetch_size`.

**Comment by denesb** [comment]
The way I see it, ScyllaDB's interpretation is much more efficient and safe (in the sense of fitting into the timeout period) while Cassandra's is more intuitive.

**Comment by nyh** [comment]
> The way I see it, ScyllaDB's interpretation is much more efficient (no need to drop over-read rows) and safe (in the sense of fitting into the timeout period) while Cassandra's is more intuitive.

I agree. One of the options we have now is to decide that we think that Scylla's behavior is the better one, and keep it. If we decide that, we should:
1. Add to our documentation an explanation of Scylla's paging behavior, and preferably also mention that it's different from Cassandra's (for the benefit of users switching from Cassandra).
2. Change he new test added in #12103 to test Scylla's behavior (expect to read a page of one item, not three), and mark the test with the "scylla_only" or "cassandra_bug" fixture, and explain in a comment that we consider Scylla's behavior to be the better or more correct one, and don't intend to be compatible with Cassandra on that issue.

**Comment by avikivity** [comment]
cqlsh paging should be decoupled from driver/scylladb page sizes. cqlsh should page screenfulls (can be just one row if the row is very long or if the screen is very short). A screenfull can be part of a driver page, or multiple driver pages.

**Comment by nyh** [comment]
> cqlsh paging should be decoupled from driver/scylladb page sizes. cqlsh should page screenfulls (can be just one row if the row is very long or if the screen is very short). A screenfull can be part of a driver page, or multiple driver pages.

I agree. This has become the focus of https://github.com/scylladb/scylladb/issues/11340, and why I opened this issue separately.
